engine/source/runtime/function/framework/world/world_manager.cpp

```diff
@@ -1,121 +1,137 @@
 #include "runtime/resource/asset_manager/asset_manager.h"
 #include "runtime/resource/config_manager/config_manager.h"
 
-#include "runtime/function/framework/level/level.h"
-
-#include <unordered_set>
-
 #include "_generated/serializer/all_serializer.h"
 
 namespace Pilot
 {
     WorldManager::~WorldManager() { clear(); }
 
-    void WorldManager::initialize() { m_pending_load_world_url = ConfigManager::getInstance().getDefaultWorldUrl(); }
+    void WorldManager::initialize()
+    {
+        m_is_world_loaded   = false;
+        m_current_world_url = ConfigManager::getInstance().getDefaultWorldUrl();
+    }
 
     void WorldManager::clear()
     {
-        m_pending_load_world_url.clear();
-        m_current_world_url.clear();
-        m_current_level_url.clear();
-
-        m_current_active_level = nullptr;
-        for (Level* level : m_levels)
+        // unload all loaded levels
+        for (auto level_pair : m_loaded_levels)
         {
-            level->clear();
-            delete level;
+            level_pair.second->unload();
         }
-        m_levels.clear();
+        m_loaded_levels.clear();
+
+        m_current_active_level.reset();
+
+        // clear world
+        m_current_world_resource.reset();
+        m_current_world_url.clear();
+        m_is_world_loaded = false;
     }
 
     void WorldManager::tick(float delta_time)
     {
-        processPendingLoadWorld();
+        if (!m_is_world_loaded)
+        {
+            loadWorld(m_current_world_url);
+        }
 
-        for (Level* level : m_levels)
+        // tick the active level
+        auto active_level = m_current_active_level.lock();
+        if (active_level)
         {
-            level->tickAll(delta_time);
+            active_level->tick(delta_time);
         }
     }
 
-    void WorldManager::processPendingLoadWorld()
+    bool WorldManager::loadWorld(const std::string& world_url)
     {
-        if (m_pending_load_world_url.empty())
-            return;
+        LOG_INFO("loading world: {}", world_url);
+        WorldRes   world_res;
+        const bool is_world_load_success = AssetManager::getInstance().loadAsset(world_url, world_res);
+        if (!is_world_load_success)
+        {
+            return false;
+        }
 
-        std::string pending_load_world_url = m_pending_load_world_url;
-        clear();
+        m_current_world_resource = std::make_shared<WorldRes>(world_res);
 
-        loadWorld(pending_load_world_url);
-    }
+        const bool is_level_load_success = loadLevel(world_res.m_default_level_url);
+        if (!is_level_load_success)
+        {
+            return false;
+        }
 
-    void WorldManager::loadWorld(const std::string& world_url)
-    {
-        WorldRes world_res;
-        AssetManager::getInstance().loadAsset(world_url, world_res);
+        // set the default level to be active level
+        auto iter = m_loaded_levels.find(world_res.m_default_level_url);
+        ASSERT(iter != m_loaded_levels.end());
```

Comment:
As we don't assert on resource error, we should LOG_FATAL here

Reply:
The logic of the above code will guarantee that the default level should be in loaded levels. This ASSERT is used for readiblity